### ISSUE TITLE
Refactor: Extract Method

### ISSUE THREAD

**Issue by markmandel** [go-lang-plugin-org/go-lang-idea-plugin#604]
I use `Extract Method` a great deal when working in other languages in IntelliJ (Java, Ruby, etc).

https://www.jetbrains.com/idea/webhelp/extract-method.html

I think it would make an awesome edition to this otherwise fantastic plugin.

**Comment by SchumacherFM** [comment]
I've just missed this feature to extract an interface from a struct with all its method receivers ...

**Comment by dlsniper** [comment]
@SchumacherFM I think you've just missed a different feature than the original poster :) 

Please open a different issue for this as you'd rather like to have a "Generate interface from struct" and the issue refers to extracting code from a method and creating another method.

We also welcome contributions and I know for sure that Alexander and Sergey will do their best to guide people thru the implementation (I don't have that much experience).

**Comment by SchumacherFM** [comment]
:laughing: thanks for pointing out. Indeed I've only searched for "extract" in the issues.

I've never coded in Java so I can only throw :moneybag: at you.

**Comment by mdwhatcott** [comment]
Would the code that implements this refactoring need to be written in Java or could it be written in Go and wrapped by Java plugin code?

**Comment by dlsniper** [comment]
@mdwhatcott I don't think it would be easy to create all the functionality the IntelliJ platform exposes in order to make this kind of bridge work. If you are not familiar with this particular functionality, just try it quickly in a Java / Android code or some other language that has this already to see what I mean (or their online help https://www.jetbrains.com/idea/help/extract-method.html)

But then again, I'm known to be wrong sometimes so I defer to Sergey or Alexander in case this is one of those cases. Guys if it's something doable / present in an open-source plugin do you have any reference for it? Maybe someone else could pick this up and contribute it to the plugin.

**Comment by mdwhatcott** [comment]
@dlsniper - In order to perform this refactoring, you have to analyze the source code to identity local variables and anticipate return types. Go already has packages for analyzing go source code (http://golang.org/pkg/go/ast/). I'm hoping I can use a go program to do the heavy lifting (analysis, insertion of new method, etc...) and then just tie that into the IDE with some java code that calls it. Would this be possible, or has someone already gone to the trouble of writing introspection code in java that understands go code?

**Comment by dlsniper** [comment]
@mdwhatcott I wasn't even considering that part as that's what the plugin is supposed to have at some point anyway, due to a host of other functionality such smarter completion for example. The goal is to have the same functionality as you'd expect from the Java engine / other languages that work on the IntelliJ Platform but it's a journey to get there :)

There was some work done by @mtoader at some point last year but for the old branch of the plugin, I can't find it right now. As for the new branch of the plugin (current version++) Sergey and Alexander can better judge this (and when it will happen).

I think I speak for them as well when I say that any contributions to the plugin are welcomed, especially in these areas, so if you (or anyone else) plan to do something, let us know and we can help out.

**Comment by zolotov** [comment]
> Would this be possible, or has someone already gone to the trouble of writing introspection code in java that understands go code?

@mdwhatcott it is possible, but it doesn't make any sense to me. Go plugin also has packages for analizing go sources, so what's the point in introducing interaction introspection between go and java?

**Comment by mdwhatcott** [comment]
Thanks @zolotov for answering my question. If there are packages for analyzing go source code than it makes no sense for me to write that in go. I am completely new to this project so I didn't know what had been done. (Also, I'm not a java developer so writing in go seemed easier to me.)

Can you point me in the direction of said analysis code?

**Comment by zolotov** [comment]
@mdwhatcott actually the most of the code of plugin is analyzing. `com.goide.psi` package contains object-representations of source code (it's AST actually but with some semantic information). You can read some docs about PSI here: https://confluence.jetbrains.com/display/IDEADEV/Developing+Custom+Language+Plugins+for+IntelliJ+IDEA#DevelopingCustomLanguagePluginsforIntelliJIDEA-ImplementingaParserandPSI and https://confluence.jetbrains.com/display/IDEADEV/IntelliJ+IDEA+Architectural+Overview#IntelliJIDEAArchitecturalOverview-PsiElements.

Also you can find examples of working with Go source code in `*Inspection` classes.

**Comment by zolotov** [comment]
Also [tutorial](https://confluence.jetbrains.com/display/IntelliJIDEA/Custom+Language+Support) could be useful

**Comment by mdwhatcott** [comment]
Thanks!

**Comment by ignatov** [comment]
Please take a look at http://www.jetbrains.org/intellij/sdk/docs/​

**Comment by mdwhatcott** [comment]
Could someone add the "refactor" label to this issue?

**Comment by akavel** [comment]
I suppose this could be used to implement the feature:

https://github.com/godoctor/godoctor (see: http://gorefactor.org/doc.html#refactoring-extract)

**Comment by mdwhatcott** [comment]
@akavel - If I understand what @zolotov wrote earlier in this thread the goal is to code the analysis and the refactoring behavior within the plugin, which is written in Java. It would be great if the plugin could make use of [godoctor](https://github.com/godoctor/godoctor), but I'm not sure if that would work.

**Comment by mdwhatcott** [comment]
I would love to be helpful here but getting bootstrapped to develop the plugin as well as learning java and the architecture of the plugin seems a bit daunting to me at the moment. I'm really hopeful someone who knows what they are doing will begin work on this feature sometime soon.

**Comment by dlsniper** [comment]
@mdwhatcott you can start with simpler things such as the issues labeled as [up for grabs](https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues?q=is%3Aissue+is%3Aopen+label%3A%22up+for+grabs%22) as that clears time from other issues (such as this ;) )
